logstash-core/src/main/java/org/logstash/instrument/metrics/FlowMetric.java

```diff
@@ -0,0 +1,124 @@
+    static final String CURRENT_KEY = "current";
+
+    public FlowMetric(final String name,
+                      final Metric<? extends Number> numeratorMetric,
+                      final Metric<? extends Number> denominatorMetric) {
+        this(System::nanoTime, name, numeratorMetric, denominatorMetric);
+    }
+
+    FlowMetric(final LongSupplier nanoTimeSupplier,
+               final String name,
+               final Metric<? extends Number> numeratorMetric,
+               final Metric<? extends Number> denominatorMetric) {
+        super(name);
+        this.nanoTimeSupplier = nanoTimeSupplier;
+        this.numeratorMetric = numeratorMetric;
+        this.denominatorMetric = denominatorMetric;
+
+        this.baseline = doCapture();
+        this.head = new AtomicReference<>(this.baseline);
+    }
+
+    public void capture() {
+        final Capture newestHead = doCapture();
+        final Capture previousHead = head.getAndSet(newestHead);
+        instant.getAndAccumulate(previousHead, (current, given) -> {
+            // keep our current value if the given one is less than ~100ms older than our newestHead
+            // this is naive and when captures happen too frequently without relief can result in
+            // our "current" window growing indefinitely, but we are shipping with a 5s cadence
+            // and shouldn't hit this edge-case in practice.
+            return (newestHead.calculateCapturePeriod(given).toMillis() > 100) ? given : current;
+        });
+    }
+
+    public Map<String, Double> getValue() {
+        final Capture headCapture = head.get();
+        if (Objects.isNull(headCapture)) {
+            return Map.of();
+        }
+
+        final Map<String, Double> rates = new HashMap<>();
+
+        final Double lifetimeRate = headCapture.calculateRate(baseline);
+
+        if (Objects.nonNull(lifetimeRate)) {
+            rates.put(LIFETIME_KEY, lifetimeRate);
+        }
+
+        final Capture instantCapture = instant.get();
+        final Double currentRate = headCapture.calculateRate(instantCapture);
+        if (Objects.nonNull(currentRate)) {
+            rates.put(CURRENT_KEY, currentRate);
+        }
+
+        return Map.copyOf(rates);
+    }
+
+    Capture doCapture() {
+        return new Capture(numeratorMetric.getValue(), denominatorMetric.getValue(), nanoTimeSupplier.getAsLong());
+    }
+
+    @Override
+    public MetricType getType() {
+        return MetricType.FLOW_RATE;
+    }
+
+    private static class Capture {
+        private final Number numerator;
+        private final Number denominator;
+
+        private final long nanoTimestamp;
+
+        public Capture(final Number numerator, final Number denominator, final long nanoTimestamp) {
+            this.numerator = numerator;
+            this.denominator = denominator;
+            this.nanoTimestamp = nanoTimestamp;
+        }
+
+        Double calculateRate(final Capture baseline) {
```

Comment:
nit: we could use `Optional` since we are returning `null` values to avoid `NPE`s.